zjit/src/codegen.rs

```diff
@@ -913,6 +914,36 @@ fn gen_new_array(
     new_array
 }
 
+/// Compile a new hash instruction
+fn gen_new_hash(
+    jit: &mut JITState,
+    asm: &mut Assembler,
+    elements: &Vec<(InsnId, InsnId)>,
+    state: &FrameState,
+) -> lir::Opnd {
+    // Save PC
+    gen_save_pc(asm, state);
+
+    asm_comment!(asm, "call rb_hash_new");
+    let cap: ::std::os::raw::c_long = (elements.len() / 2).try_into().expect("Unable to fit length of elements into c_long");
+    let new_hash = asm.ccall(
+        rb_hash_new_with_size as *const u8,
+        vec![lir::Opnd::Imm(cap)],
+    );
+
+    for (key_id, val_id) in elements.iter() {
+        let key = jit.get_opnd(*key_id).unwrap();
+        let val = jit.get_opnd(*val_id).unwrap();
+        asm_comment!(asm, "call rb_hash_aset");
+        asm.ccall(
+            rb_hash_aset as *const u8,
```

Comment:
I think this can call arbitrary user code so we have to save more state than just the PC. I'm not 100% sure what, in this case. I'll get back to you

Reply:
I think we need to write all the locals and stack and SP to the frame, unfortunately. See for example the beginning of gen_send_without_block. I would probably copy that code